NeoML/include/NeoML/TraditionalML/FloatVector.h

```diff
@@ -163,58 +195,65 @@ inline CFloatVector::TIterator CFloatVector::end()
 	return TIterator( CopyOnWrite() + Size() );
 }
 
-// The dot product of two vectors
-inline double DotProduct( const CFloatVector& vector1, const CFloatVector& vector2 )
-{
-	NeoPresume( vector1.Size() == vector2.Size() );
-	double sum = 0;
-
-	const int size = vector1.Size();
-	const float* operand1 = vector1.GetPtr();
-	const float* operand2 = vector2.GetPtr();
-
-	for( int i = 0; i < size; i++ ) {
-		sum += static_cast<double>( operand1[i] ) * operand2[i];
-	}
-	return sum;
-}
-
 // The dot product of two vectors
 inline double DotProduct( const CSparseFloatVectorDesc& vector1, const CSparseFloatVectorDesc& vector2 )
 {
-	int i = 0;
-	int j = 0;
 	double sum = 0;
-	while( i < vector1.Size && j < vector2.Size ) {
-		if( vector1.Indexes[i] == vector2.Indexes[j] ) {
-			sum += static_cast<double>( vector1.Values[i] ) * vector2.Values[j];
-			i++;
-			j++;
+
+	auto denseBySparse = []( const CSparseFloatVectorDesc& dense, const CSparseFloatVectorDesc& sparse, double& sum ) {
+		if( sparse.Indexes[0] >= dense.Size ) {
```

Comment:
for( int i = 0; i < sparse.Size && sparse.Indexes[i] < dense.Size; i++ ) {
		sum += static_cast<double>( sparse.Values[i] ) * dense.Values[sparse.Indexes[i]];
	}